File: design-patterns/go/decorator/ui_element/ui_decorator.go
// ui_decorator.go

package main

type UIDecorator struct {
	uiElement UIElement
}

func NewUIDecorator(uiElement UIElement) (uiDecorator *UIDecorator) {
	uiDecorator = &UIDecorator{}
	uiDecorator.uiElement = uiElement
	return
}

func (uiDecorator *UIDecorator) Draw() {
	uiDecorator.uiElement.Draw()
}

File: design-patterns/ts/decorator/ui-element/ui-decorator.ts
// ui-decorator.ts

import UIElement from "./ui-element";

abstract class UIDecorator implements UIElement {
    protected readonly uiElement: UIElement;

    protected constructor(uiElement: UIElement) {
        this.uiElement = uiElement;
    }

    draw(): void {
        this.uiElement.draw();
    }
}

export default UIDecorator;